src/Aspire.Hosting/Codespaces/CodespacesUrlRewriter.cs

```diff
@@ -0,0 +1,89 @@
+        {
+            logger.LogTrace("Not running in Codespaces, skipping URL rewriting.");
+            return;
+        }
+
+        var gitHubCodespacesPortForwardingDomain = GetRequiredCodespacesConfigurationValue(GitHubCodespacesPortForwardingDomain);
+        var codespaceName = GetRequiredCodespacesConfigurationValue(CodespaceNameEnvironmentVariable);
+
+        do
+        {
+            try
+            {
+                var resourceEvents = resourceNotificationService.WatchAsync(stoppingToken);
+
+                await foreach (var resourceEvent in resourceEvents.ConfigureAwait(false))
+                {
+                    Dictionary<UrlSnapshot, UrlSnapshot>? remappedUrls = null;
+
+                    foreach (var originalUrlSnapshot in resourceEvent.Snapshot.Urls)
+                    {
+                        var uri = new Uri(originalUrlSnapshot.Url);
+
+                        if (!originalUrlSnapshot.IsInternal && (uri.Scheme == "http" || uri.Scheme == "https") && uri.Host == "localhost")
+                        {
+                            remappedUrls ??= new();
+
+                            var newUrlSnapshot = originalUrlSnapshot with
+                            {
+                                // The format of GitHub Codespaces URLs comprises the codespace
+                                // name (from the CODESPACE_NAME environment variable, the port,
+                                // and the port forwarding domain (via GITHUB_CODESPACES_PORT_FORWARDING_DOMAIN
+                                // which is typically ".app.github.dev". The VSCode instance is typically
+                                // hosted at codespacename.github.dev whereas the forwarded ports
+                                // would be at codespacename-port.app.github.dev.
+                                Url = $"{uri.Scheme}://{codespaceName}-{uri.Port}.{gitHubCodespacesPortForwardingDomain}{uri.AbsolutePath}"
+                            };
+
+                            remappedUrls.Add(originalUrlSnapshot, newUrlSnapshot);
+                        }
+                    }
+
+                    if (remappedUrls is not null)
+                    {
+                        var transformedUrls = from originalUrl in resourceEvent.Snapshot.Urls
+                                              select remappedUrls.TryGetValue(originalUrl, out var remappedUrl) ? remappedUrl : originalUrl;
+
+                        await resourceNotificationService.PublishUpdateAsync(resourceEvent.Resource, resourceEvent.ResourceId, s => s with
+                        {
+                            Urls = transformedUrls.ToImmutableArray()
+                        }).ConfigureAwait(false);
+                    }
+                }
+            }
+            catch (Exception ex) when (!stoppingToken.IsCancellationRequested)
+            {
+                // When debugging sometimes we'll get cancelled here but we don't want
```

Comment:
The delay was removed. Intentional?

Reply:
Yeah, I looked at how this was done in other areas and the delay isn't really necessary. The Watch is a natural throttling issue. The reason why we catch and retry is mostly for debuggability. I've found that in certain cases if I am debugging these loops the cancellation can occur and the background service stops.

When I tracked the cancellation back at one point it appeared to be the dashboard hanging up after a timeout. In those cases we just want to restart watching.